File: palettier-core/src/main/java/be/asmolabs/palettier/core/image/ImagePalette.java
package be.asmolabs.palettier.core.image;

import be.asmolabs.palettier.core.color.Colors;
import be.asmolabs.palettier.core.color.Lab;
import be.asmolabs.palettier.core.color.Rgb;
import java.awt.image.BufferedImage;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import javax.imageio.ImageIO;

/**
 * Les teintes reellement presentes sur une photo.
 *
 * <p>Sert a fonder un plan sur ce qu'on voit plutot que sur ce qu'on imagine. Un modele
 * de langage a qui l'on montre une photo de carnation propose volontiers des couleurs
 * de manuel ; lui donner les teintes mesurees l'oblige a travailler sur la piece qu'il
 * a sous les yeux.</p>
 *
 * <p>Le regroupement se fait par distance perceptuelle en L*a*b*, parce qu'un ecart de
 * dix unites y signifie la meme chose dans les ombres et dans les clairs, ce qui n'est
 * pas le cas en RVB. Les representants, eux, restent des moyennes en lumiere lineaire :
 * moyenner du sRGB assombrit.</p>
 */
public final class ImagePalette {

    /** Cote maximal avant analyse : la teinte dominante ne demande pas de resolution. */
    private static final int ANALYSIS_EDGE = 160;

    /** Nombre d'iterations : au-dela, les groupes ne bougent plus. */
    private static final int ITERATIONS = 12;

    private ImagePalette() {
    }

    /**
     * Une teinte dominante.
     *
     * @param share part de l'image occupee, de 0 a 1
     */
    public record DominantColour(Rgb color, double share) {
    }

    /**
     * Extrait les teintes dominantes d'une image.
     *
     * @param count nombre de teintes souhaitees
     * @return les teintes, de la plus presente a la moins presente
     */
    public static List<DominantColour> dominant(byte[] image, int count) {
        double[][] pixels = linearPixels(image);
        if (pixels.length == 0 || count < 1) {
            return List.of();
        }

        int groups = Math.min(count, pixels.length);
        double[][] centres = seed(pixels, groups);
        int[] assignment = new int[pixels.length];

        for (int pass = 0; pass < ITERATIONS; pass++) {
            Lab[] centreLab = labOf(centres);
            boolean moved = false;

            for (int i = 0; i < pixels.length; i++) {
                int best = nearest(pixels[i], centreLab);
                if (assignment[i] != best) {
                    assignment[i] = best;
                    moved = true;
                }
            }
            recentre(pixels, assignment, centres);
            if (!moved) {
                break;
            }
        }

        return summarise(pixels, assignment, centres);
    }

    // --- Lecture -----------------------------------------------------------

    /** Pixels en lumiere lineaire : c'est la qu'une moyenne a un sens. */
    private static double[][] linearPixels(byte[] image) {
        try {
            BufferedImage source = ImageIO.read(new ByteArrayInputStream(image));
            if (source == null) {
                throw new IllegalArgumentException("Format d'image non reconnu");
            }
            int step = Math.max(1, Math.max(source.getWidth(), source.getHeight()) / ANALYSIS_EDGE);

            List<double[]> pixels = new ArrayList<>();
            for (int y = 0; y < source.getHeight(); y += step) {
                for (int x = 0; x < source.getWidth(); x += step) {
                    int packed = source.getRGB(x, y);
                    pixels.add(new double[]{
                            Colors.srgbToLinear(((packed >> 16) & 0xFF) / 255.0),
                            Colors.srgbToLinear(((packed >> 8) & 0xFF) / 255.0),
                            Colors.srgbToLinear((packed & 0xFF) / 255.0)});
                }
            }
            return pixels.toArray(double[][]::new);
        } catch (IOException e) {
            throw new UncheckedIOException("Lecture de l'image impossible", e);
        }
    }

    // --- Regroupement ------------------------------------------------------

    /**
     * Points de depart etales : on prend le premier pixel, puis a chaque fois le plus
     * eloigne de ceux deja retenus. Deterministe, contrairement a un tirage au sort, ce
     * qui evite qu'une meme photo donne deux plans differents.
     */
    private static double[][] seed(double[][] pixels, int groups) {
        double[][] centres = new double[groups][];
        centres[0] = pixels[0].clone();

        for (int taken = 1; taken < groups; taken++) {
            Lab[] chosen = labOf(java.util.Arrays.copyOf(centres, taken));
            int farthest = 0;
            double worst = -1;
            for (int i = 0; i < pixels.length; i++) {
                Lab lab = toLab(pixels[i]);
                double nearest = Double.MAX_VALUE;
                for (Lab centre : chosen) {
                    nearest = Math.min(nearest, Colors.deltaE2000(lab, centre));
                }
                if (nearest > worst) {
                    worst = nearest;
                    farthest = i;
                }
            }
            centres[taken] = pixels[farthest].clone();
        }
        return centres;
    }

    private static int nearest(double[] pixel, Lab[] centres) {
        Lab lab = toLab(pixel);
        int best = 0;
        double shortest = Double.MAX_VALUE;
        for (int i = 0; i < centres.length; i++) {
            double distance = Colors.deltaE2000(lab, centres[i]);
            if (distance < shortest) {
                shortest = distance;
                best = i;
            }
        }
        return best;
    }

    private static void recentre(double[][] pixels, int[] assignment, double[][] centres) {
        double[][] sums = new double[centres.length][3];
        int[] counts = new int[centres.length];

        for (int i = 0; i < pixels.length; i++) {
            int group = assignment[i];
            counts[group]++;
            for (int channel = 0; channel < 3; channel++) {
                sums[group][channel] += pixels[i][channel];
            }
        }
        for (int group = 0; group < centres.length; group++) {
            if (counts[group] > 0) {
                for (int channel = 0; channel < 3; channel++) {
                    centres[group][channel] = sums[group][channel] / counts[group];
                }
            }
        }
    }

    private static List<DominantColour> summarise(double[][] pixels, int[] assignment, double[][] centres) {
        int[] counts = new int[centres.length];
        for (int group : assignment) {
            counts[group]++;
        }

        List<DominantColour> found = new ArrayList<>();
        for (int group = 0; group < centres.length; group++) {
            if (counts[group] > 0) {
                found.add(new DominantColour(toRgb(centres[group]),
                        (double) counts[group] / pixels.length));
            }
        }
        found.sort(Comparator.comparingDouble(DominantColour::share).reversed());
        return List.copyOf(found);
    }

    // --- Conversions -------------------------------------------------------

    private static Lab[] labOf(double[][] linear) {
        Lab[] labs = new Lab[linear.length];
        for (int i = 0; i < linear.length; i++) {
            labs[i] = toLab(linear[i]);
        }
        return labs;
    }

    private static Lab toLab(double[] linear) {
        return Colors.toLab(toRgb(linear));
    }

    private static Rgb toRgb(double[] linear) {
        return new Rgb(Colors.linearToSrgb(linear[0]),
                Colors.linearToSrgb(linear[1]),
                Colors.linearToSrgb(linear[2]));
    }
}

File: palettier-kmp/core-domain/src/commonMain/kotlin/be/asmolabs/palettier/domain/image/ImagePalette.kt
package be.asmolabs.palettier.domain.image

import be.asmolabs.palettier.domain.color.Lab
import be.asmolabs.palettier.domain.color.Rgb
import be.asmolabs.palettier.domain.color.deltaE2000
import be.asmolabs.palettier.domain.color.linearToSrgb
import be.asmolabs.palettier.domain.color.srgbToLinear
import be.asmolabs.palettier.domain.color.toLab
import be.asmolabs.palettier.domain.project.DominantColour

/**
 * Une image reduite a ce que le domaine en a besoin : des pixels.
 *
 * <p>Decoder un JPEG n'a rien de commun d'une plateforme a l'autre, mais compter des
 * teintes, si. La frontiere passe donc ici : la plateforme rend un PixelMap, le domaine
 * fait le reste -- et ne le fait qu'une fois.</p>
 *
 * @param argb un entier par pixel, au format 0xAARRGGBB, en lecture ligne par ligne
 */
data class PixelMap(val width: Int, val height: Int, val argb: IntArray) {
    init {
        require(width >= 0 && height >= 0) { "Dimensions negatives" }
        require(argb.size == width * height) { "${argb.size} pixels pour ${width}x$height" }
    }

    val isEmpty: Boolean get() = argb.isEmpty()

    // IntArray casse l'egalite structurelle d'une data class : deux images sont les memes
    // si elles ont les memes pixels, pas la meme reference.
    override fun equals(other: Any?) = other is PixelMap &&
        width == other.width && height == other.height && argb.contentEquals(other.argb)

    override fun hashCode() = (width * 31 + height) * 31 + argb.contentHashCode()
}

/**
 * Les teintes reellement presentes sur une photo.
 *
 * <p>Sert a fonder un plan sur ce qu'on voit plutot que sur ce qu'on imagine. Un modele
 * de langage a qui l'on montre une photo de carnation propose volontiers des couleurs de
 * manuel ; lui donner les teintes mesurees l'oblige a travailler sur la piece qu'il a
 * sous les yeux.</p>
 *
 * <p>Le regroupement se fait par distance perceptuelle en L*a*b*, parce qu'un ecart de
 * dix unites y signifie la meme chose dans les ombres et dans les clairs, ce qui n'est
 * pas le cas en RVB. Les representants, eux, restent des moyennes en lumiere lineaire :
 * moyenner du sRGB assombrit.</p>
 */
object ImagePalette {

    /** Cote maximal avant analyse : la teinte dominante ne demande pas de resolution. */
    private const val ANALYSIS_EDGE = 160

    /** Nombre d'iterations : au-dela, les groupes ne bougent plus. */
    private const val ITERATIONS = 12

    /**
     * Extrait les teintes dominantes d'une image.
     *
     * @param count nombre de teintes souhaitees
     * @return les teintes, de la plus presente a la moins presente
     */
    fun dominant(image: PixelMap, count: Int): List<DominantColour> {
        val pixels = linearPixels(image)
        if (pixels.isEmpty() || count < 1) return emptyList()

        val groups = minOf(count, pixels.size)
        val centres = seed(pixels, groups)
        val assignment = IntArray(pixels.size)

        repeat(ITERATIONS) {
            val centreLab = centres.map { it.toLabPoint() }
            var moved = false

            for (i in pixels.indices) {
                val best = nearest(pixels[i], centreLab)
                if (assignment[i] != best) {
                    assignment[i] = best
                    moved = true
                }
            }
            recentre(pixels, assignment, centres)
            if (!moved) return@repeat
        }

        return summarise(pixels, assignment, centres)
    }

    /** Pixels en lumiere lineaire : c'est la qu'une moyenne a un sens. */
    private fun linearPixels(image: PixelMap): List<DoubleArray> {
        if (image.isEmpty) return emptyList()
        val step = maxOf(1, maxOf(image.width, image.height) / ANALYSIS_EDGE)

        val pixels = ArrayList<DoubleArray>()
        var y = 0
        while (y < image.height) {
            var x = 0
            while (x < image.width) {
                val packed = image.argb[y * image.width + x]
                pixels += doubleArrayOf(
                    srgbToLinear(((packed shr 16) and 0xFF) / 255.0),
                    srgbToLinear(((packed shr 8) and 0xFF) / 255.0),
                    srgbToLinear((packed and 0xFF) / 255.0),
                )
                x += step
            }
            y += step
        }
        return pixels
    }

    /**
     * Points de depart etales : le premier pixel, puis a chaque fois le plus eloigne de
     * ceux deja retenus. Deterministe, contrairement a un tirage au sort -- ce qui evite
     * qu'une meme photo donne deux plans differents.
     */
    private fun seed(pixels: List<DoubleArray>, groups: Int): MutableList<DoubleArray> {
        val centres = ArrayList<DoubleArray>(groups)
        centres += pixels[0].copyOf()

        while (centres.size < groups) {
            val chosen = centres.map { it.toLabPoint() }
            var farthest = 0
            var worst = -1.0
            for (i in pixels.indices) {
                val lab = pixels[i].toLabPoint()
                val nearest = chosen.minOf { deltaE2000(lab, it) }
                if (nearest > worst) {
                    worst = nearest
                    farthest = i
                }
            }
            centres += pixels[farthest].copyOf()
        }
        return centres
    }

    private fun nearest(pixel: DoubleArray, centres: List<Lab>): Int {
        val lab = pixel.toLabPoint()
        var best = 0
        var shortest = Double.MAX_VALUE
        for (i in centres.indices) {
            val distance = deltaE2000(lab, centres[i])
            if (distance < shortest) {
                shortest = distance
                best = i
            }
        }
        return best
    }

    private fun recentre(pixels: List<DoubleArray>, assignment: IntArray, centres: MutableList<DoubleArray>) {
        val sums = Array(centres.size) { DoubleArray(3) }
        val counts = IntArray(centres.size)

        for (i in pixels.indices) {
            val group = assignment[i]
            counts[group]++
            for (channel in 0..2) sums[group][channel] += pixels[i][channel]
        }
        for (group in centres.indices) {
            if (counts[group] > 0) {
                for (channel in 0..2) centres[group][channel] = sums[group][channel] / counts[group]
            }
        }
    }

    private fun summarise(
        pixels: List<DoubleArray>,
        assignment: IntArray,
        centres: List<DoubleArray>,
    ): List<DominantColour> {
        val counts = IntArray(centres.size)
        for (group in assignment) counts[group]++

        return centres.indices
            .filter { counts[it] > 0 }
            .map { DominantColour(centres[it].toRgb(), counts[it].toDouble() / pixels.size) }
            .sortedByDescending { it.share }
    }

    private fun DoubleArray.toRgb() = Rgb(linearToSrgb(this[0]), linearToSrgb(this[1]), linearToSrgb(this[2]))
    private fun DoubleArray.toLabPoint(): Lab = toRgb().toLab()
}
